### PR TITLE
Add short traceback to generated errors in developer mode

### PR DESCRIPTION
**Pull request by EmilyBourne**
Pyccel has a `--developer-mode` option which does nothing. This PR modifies the error messages if Pyccel is run in developer mode. Instead of just printing the error, it also prints the last 5 calls on the stack.

E.g. The following file:
```
import numpy
import numpy

from pyccel.decorators import types

@unknown
@types('int', 'int')
def func(n):
    return n
```
when compiled with `pyccel test.py` produces the following output:
```
pyccel:
 |warning [semantic]: warn_central.py| Duplicated import  (numpy)
 |warning [semantic]: warn_central.py| Decorator(s) not used (unknown)
 |warning [semantic]: warn_central.py| The number of arguments in the function {} ({}) does not match the number
                        of types in decorator/header ({}).'.format(name ,args_number, len(hd.dtypes)) ((<pyccel.ast.core.Argument object at 0x7f5f07c9e0a0>,))
```
but when run with `pyccel test.py --developer-mode`, it produces the following output:
```
pyccel:
   File "/home/emily/Code/pyccel/pyccel/parser/semantic.py", line 837, in <listcomp>
    ls = [self._visit(i, **settings) for i in expr.body]
  File "/home/emily/Code/pyccel/pyccel/parser/semantic.py", line 801, in _visit
    obj = getattr(self, annotation_method)(expr, **settings)
  File "/home/emily/Code/pyccel/pyccel/parser/semantic.py", line 2881, in _visit_Import
    _insert_obj('variables', source_target, imports)
  File "/home/emily/Code/pyccel/pyccel/parser/semantic.py", line 2864, in _insert_obj
    errors.report(FOUND_DUPLICATED_IMPORT,
  File "/home/emily/Code/pyccel/pyccel/errors/errors.py", line 254, in report
    traceback = ''.join(tb.format_stack(limit=5))
|warning [semantic]: warn_central.py| Duplicated import  (numpy)
   File "/home/emily/Code/pyccel/pyccel/parser/semantic.py", line 837, in _visit_CodeBlock
    ls = [self._visit(i, **settings) for i in expr.body]
  File "/home/emily/Code/pyccel/pyccel/parser/semantic.py", line 837, in <listcomp>
    ls = [self._visit(i, **settings) for i in expr.body]
  File "/home/emily/Code/pyccel/pyccel/parser/semantic.py", line 801, in _visit
    obj = getattr(self, annotation_method)(expr, **settings)
  File "/home/emily/Code/pyccel/pyccel/parser/semantic.py", line 2372, in _visit_FunctionDef
    errors.report(UNUSED_DECORATORS, symbol=', '.join(not_used), severity='warning')
  File "/home/emily/Code/pyccel/pyccel/errors/errors.py", line 254, in report
    traceback = ''.join(tb.format_stack(limit=5))
|warning [semantic]: warn_central.py| Decorator(s) not used (unknown)
   File "/home/emily/Code/pyccel/pyccel/parser/semantic.py", line 837, in _visit_CodeBlock
    ls = [self._visit(i, **settings) for i in expr.body]
  File "/home/emily/Code/pyccel/pyccel/parser/semantic.py", line 837, in <listcomp>
    ls = [self._visit(i, **settings) for i in expr.body]
  File "/home/emily/Code/pyccel/pyccel/parser/semantic.py", line 801, in _visit
    obj = getattr(self, annotation_method)(expr, **settings)
  File "/home/emily/Code/pyccel/pyccel/parser/semantic.py", line 2397, in _visit_FunctionDef
    errors.report(msg, symbol=expr.arguments, severity='warning')
  File "/home/emily/Code/pyccel/pyccel/errors/errors.py", line 254, in report
    traceback = ''.join(tb.format_stack(limit=5))
|warning [semantic]: warn_central.py| The number of arguments in the function {} ({}) does not match the number
                        of types in decorator/header ({}).'.format(name ,args_number, len(hd.dtypes)) ((<pyccel.ast.core.Argument object at 0x7f2a42160070>,))
```

**Commit Summary**

- Save `ErrorsMode` object not `ErrorsMode().value` in `Errors()` so initialisation order doesn't matter
- Collect traceback when reporting errors if in developer mode
- Print traceback at start of ErrorInfo when it is provided
- Add `developer_mode` to epyccel kwargs
- [BUG_FIX] Use a multiline string for error message, not a single line string with a line continuation character
- Add `--developer-mode` to pytest options in the tests/epyccel folder to help with debugging
- Use `repr` instead of `str` to print symbol when printing errors in developer mode to get as much information as possible

### CODE REVIEW

File: pyccel/epyccel.py [hunk 1]
```diff
@@ -221,6 +221,11 @@ def epyccel( python_function_or_module, **kwargs ):
     comm  = kwargs.pop('comm', None)
     root  = kwargs.pop('root', 0)
     bcast = kwargs.pop('bcast', True)
+    if kwargs.pop('developer_mode', None):
+        # this will initialize the singleton ErrorsMode
```
Comment (mostafaMamoni) [inline pyccel/epyccel.py:225]: Great idea, just a small typo: this -> This.

**Review by mostafaMamoni — APPROVED** [review]

File: pyccel/errors/errors.py [hunk 2]
```diff
@@ -84,7 +85,8 @@ def __init__(self, *, stage, filename,
                  severity=None,
                  message='',
                  symbol=None,
-                 blocker=False):
+                 blocker=False,
+                 traceback=None):
```
Comment (MenoIy) [inline pyccel/errors/errors.py:89]: won't traceback default value  as ``` '' ```  better than None, so you can avoid check here ?
```   'traceback': '{}\n'.format(self.traceback) if self.traceback else '' ```

Comment (EmilyBourne) [reply pyccel/errors/errors.py:89]: I'm not sure if this helps as I don't want to add the line break if there is no traceback

File: pyccel/epyccel.py (hunk 1 above)
Comment (EmilyBourne) [reply pyccel/epyccel.py]: Haha, that's what happens when you copy/paste code. Fixed in 1d3c907

File: pyccel/errors/errors.py (hunk 2 above)
Comment (EmilyBourne) [reply pyccel/errors/errors.py:89]: Actually the `\n` doesn't need to be there, but we still need to pass something from `Errors.report`

**Review by MenoIy — APPROVED** [review]

**Review by nhamidn — APPROVED** [review]

**Review by yguclu — APPROVED** [review]
Very nice and clean! This is very useful for developers.

**Comment by yguclu** [comment]
Provide more information in error messages when Pyccel is run in developer mode using the option `--developer-mode`. Instead of just printing the error, it also prints the last 5 calls on the stack.

**Commit Summary**

- Save `ErrorsMode` object not `ErrorsMode().value` in `Errors()` so initialisation order doesn't matter
- Collect traceback when reporting errors if in developer mode
- Print traceback at start of ErrorInfo when it is provided
- Add `developer_mode` to epyccel kwargs
- [BUG_FIX] Use a multiline string for error message, not a single line string with a line continuation character
- Add `--developer-mode` to pytest options in the tests/epyccel folder to help with debugging
- Use `repr` instead of `str` to print symbol when printing errors in developer mode to get as much information as possible

**Pull request merged**